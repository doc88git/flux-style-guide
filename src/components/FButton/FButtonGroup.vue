<template>
  <div class="f-button-group">
    <f-button
      v-for="opt in options"
      :key="opt.value"
      @click="change(opt.value)"
      :class="{
        ...classes,
        'f-button-group__tab--selected': tab && opt.value === selected
      }"
      v-bind="getAttrs(opt.value)"
    >
      {{ opt.label }}
    </f-button>
  </div>
</template>

<script>
import FButton from './FButton'

export default {
  name: 'f-button-group',
  components: {
    FButton
  },
  props: {
    options: {
      type: Array,
      required: true
    },
    outline: Boolean,
    tab: Boolean,
    default: [String, Number],
    size: {
      type: String,
      default: ''
    }
  },
  data: () => ({
    selected: null
  }),
  computed: {
    isDefault() {
      return !this.outline && !this.tab
    },
    classes() {
      return {
        'f-button-group__tab': this.tab
      }
    },
    btnOptions () {
      return {
        flat: this.tab,
        small: this.size === 'small',
        bigger: this.size === 'bigger'
      }
    }
  },
  methods: {
    change(value) {
      this.selected = value
      this.$emit('change', this.selected)
    },
    getAttrs(id) {
      return this.tab || this.outline && id === this.selected
        ? { ...this.btnOptions, outline: false }
        : this.isDefault
        ? { ...this.btnOptions, outline: id === this.selected }
        : { ...this.btnOptions, outline: true }
    },
    isSelected(id) {
      return id === this.selected
    }
  },
  mounted() {
    if (this.default) this.change(this.default)
  }
}
</script>

<style lang="scss" scoped>
.f-button-group {
  display: flex;
  &__tab {
    border-radius: 0;
    color: var(--color-gray);
    &--selected {
      color: var(--color-primary);
      border-bottom-width: 1px;
      border-color: var(--color-primary);
    }
  }
}
</style>
