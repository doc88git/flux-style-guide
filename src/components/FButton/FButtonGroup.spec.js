import { shallowMount } from '@vue/test-utils'
import FButtonGroup from './FButtonGroup'

const tabOptions = [
  { label: 'Tab 1', value: 1 },
  { label: 'Tab 2', value: 2 },
  { label: 'Tab 3', value: 3 }
] 

const WRAPPER_PROPS = {
  options: tabOptions,
  size: 'small',
  default: 2
}


describe('FButton Tests', () => {
  let WRAPPER

  beforeEach(() => WRAPPER = shallowMount(FButtonGroup, { propsData: WRAPPER_PROPS }))

  test('Componente existe', () => {
    expect(WRAPPER.exists()).toBe(true)
  })

  test('Componente não alterado', () => {
    expect(WRAPPER.html()).toMatchSnapshot()
  })

  describe('Teste de computeds', () => {
    describe('Teste da computed isDefault', () => {
      test('Computed isDefault quanto outline: true', () => {
        WRAPPER.setProps({ outline: true })
        expect(WRAPPER.vm.isDefault).toBe(false)
      })

      test('Computed isDefault quando tab: true', () => {
        WRAPPER.setProps({ tab: true })
        expect(WRAPPER.vm.isDefault).toBe(false)
      })

      test('Computed isDefault quando outline: true e tab: false', () => {
        WRAPPER.setProps({ outline: true , tab: false })
        expect(WRAPPER.vm.isDefault).toBe(false)
      })

      test('Computed isDefault quando outline: true e tab: true', () => {
        WRAPPER.setProps({ outline: true , tab: true })
        expect(WRAPPER.vm.isDefault).toBe(false)
      })

      test('Computed isDefault quando outline false e tab: false', () => {
        WRAPPER.setProps({ outline: false, tab: false })
        expect(WRAPPER.vm.isDefault).toBe(true)
      })
    })

    describe('Teste da computed classes', () => {
      test('Computed classes quando tab: false', () => {
        const objExpect = { 'f-button-group__tab': true }
        WRAPPER.setProps({ tab: true })
        expect(WRAPPER.vm.classes).toStrictEqual(objExpect)
      })

      test('Computed classes quando tab: true', () => {
        const objExpect = { 'f-button-group__tab': false }
        WRAPPER.setProps({ tab: false })
        expect(WRAPPER.vm.classes).toStrictEqual(objExpect)
      })
    })

    describe('Teste da computed btnOptions', () => {
      test('Computed btnOptions quando tab: true', () => {
        WRAPPER.setProps({ tab: true })
        expect(WRAPPER.vm.btnOptions).toEqual(expect.objectContaining({ flat: true }))
      })

      test('Computed btnOptions quando size: "small"', () => {
        const objExpect = { small: true, bigger: false }
        WRAPPER.setProps({ size: 'small' })
        expect(WRAPPER.vm.btnOptions).toEqual(expect.objectContaining(objExpect))
      })

      test('Computed btnOptions quando size: "bigger"', () => {
        const objExpect = { small: false, bigger: true }
        WRAPPER.setProps({ size: 'bigger' })
        expect(WRAPPER.vm.btnOptions).toEqual(expect.objectContaining(objExpect))
      })
    })
  })

  describe('Teste de métodos', () => {
    test('Teste do método change', async () => {
      WRAPPER.vm.change(2)
      expect(WRAPPER.vm.selected).toBe(2)

      await WRAPPER.vm.$nextTick()
      expect(WRAPPER.emitted()['change'][0][0]).toBe(2)
    })

    describe.skip('Teste do método getAttrs', () => {
      test('getAttrs quando prop tab: true', () => {
        const objExpect = { outline: false }
        WRAPPER.setProps({ tab: true })
        expect(WRAPPER.vm.getAttrs(0)).toBe(expect.objectContaining(objExpect))
      })

      test('getAttrs quando prop outline: true e id === selected', () => {
        const objExpect = { outline: false }
        WRAPPER.setProps({ outline: true, tab: false })

        expect(WRAPPER.vm.getAttrs(2)).toBe(expect.objectContaining(objExpect))
      })

      test('getAttrs quando prop outline: true e id !== selected', () => {
        const objExpect = { outline: true }
        WRAPPER.setProps({ outline: true, tab: false })

        expect(WRAPPER.vm.getAttrs(1)).toBe(expect.objectContaining(objExpect))
      })


      test('getAttrs quando isDefault: true e id === selected', () => {
        const objExpect = { outline: true }
        WRAPPER.setProps({ outline: false, tab: false })

        expect(WRAPPER.vm.getAttrs(2)).toBe(expect.objectContaining(objExpect))
      })

      test('getAttrs quando isDefault: true e id !== selected', () => {
        const objExpect = { outline: false }
        WRAPPER.setProps({ outline: false, tab: false })

        expect(WRAPPER.vm.getAttrs(3)).toBe(expect.objectContaining(objExpect))
      })
    })

    describe('Teste do método isSelected', () => {
      test('Método isSelected com argumento "2" e data selected: 2', () => {
        expect(WRAPPER.vm.isSelected(2)).toBe(true)
      })

      test('Método isSelected com argumento "1" e data selected: 2', () => {
        expect(WRAPPER.vm.isSelected(1)).toBe(false)
      })
    })
  })

  describe('Teste do método mounted', () => {
    test('Método mounted quando prop default: 1', () => {
      const change = jest.fn()
      const propsData = { ...WRAPPER_PROPS, default: 1 }
      const testWrapper = shallowMount(FButtonGroup, {
        methods: { change },
        propsData
      })

      expect(change).toHaveBeenCalledWith(1)
    })

    test('Método mounted quando prop default: 1', () => {
      const change = jest.fn()
      const propsData = { ...WRAPPER_PROPS, default: null }
      const testWrapper = shallowMount(FButtonGroup, {
        methods: { change },
        propsData
      })

      expect(change).not.toHaveBeenCalled()
    })
  })
})