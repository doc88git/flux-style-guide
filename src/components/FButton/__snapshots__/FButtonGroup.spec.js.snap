// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FButton Tests Componente n√£o alterado 1`] = `
"<div class=\\"f-button-group\\">
  <f-button-stub small=\\"true\\" color=\\"\\" radius=\\"true\\" class=\\"\\">
    Tab 1
  </f-button-stub>
  <f-button-stub small=\\"true\\" color=\\"\\" radius=\\"true\\" class=\\"\\" outline=\\"true\\">
    Tab 2
  </f-button-stub>
  <f-button-stub small=\\"true\\" color=\\"\\" radius=\\"true\\" class=\\"\\">
    Tab 3
  </f-button-stub>
</div>"
`;
